In [1]:
import sys, os
sys.path.insert(0, os.path.abspath('/home/ElasticNotebook'))
%load_ext ElasticNotebook
from elastic.core.common.pandas import compare_df
import pickle
import cudf

Enabled rmm statistics


In [2]:
# %load_ext cudf.pandas
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load
import numpy as np # linear algebra
import os
# STEFANOS: Conditionally import Modin Pandas
if "IREWR_WITH_MODIN" in os.environ and os.environ["IREWR_WITH_MODIN"] == "True":
    # STEFANOS: Import Modin Pandas
    import os
    os.environ["MODIN_ENGINE"] = "ray"
    import ray
    ray.init(num_cpus=int(os.environ['MODIN_CPUS']), runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})
    import modin.pandas as pd
else:
    # STEFANOS: Import regular Pandas
    import pandas as pd
import matplotlib.pyplot as plt # ploting graphs and chart 
from datetime import timedelta # for time formatting
import time
# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/home/dias-benchmarks/notebooks/sanket7994/imdb-dataset-eda-project/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/home/dias-benchmarks/notebooks/sanket7994/imdb-dataset-eda-project/input/imdb-official-movies-dataset/title-metadata.tsv
/home/dias-benchmarks/notebooks/sanket7994/imdb-dataset-eda-project/input/imdb-official-movies-dataset/title-crew.tsv
/home/dias-benchmarks/notebooks/sanket7994/imdb-dataset-eda-project/input/imdb-official-movies-dataset/title-ratings.tsv


In [3]:
%%time
### cell 0 ###

df = pd.read_table('/home/dias-benchmarks/notebooks/sanket7994/imdb-dataset-eda-project//input/imdb-official-movies-dataset/title-ratings.tsv', low_memory=False)
df2 = pd.read_table('/home/dias-benchmarks/notebooks/sanket7994/imdb-dataset-eda-project//input/imdb-official-movies-dataset/title-metadata.tsv',low_memory=False)
if "IREWR_LESS_REPLICATION" in os.environ and os.environ["IREWR_LESS_REPLICATION"] == "True":
    df2 = df2[:1_500_000]

CPU times: user 21.3 s, sys: 2.12 s, total: 23.4 s
Wall time: 23.4 s


In [ ]:
cudf_df2 = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['tconst'] = df2['tconst']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['titleType'] = df2['titleType']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['primaryTitle'] = df2['primaryTitle']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['originalTitle'] = df2['originalTitle']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['isAdult'] = df2['isAdult']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['startYear'] = df2['startYear']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['endYear'] = df2['endYear']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['runtimeMinutes'] = df2['runtimeMinutes']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df2['genres'] = df2['genres']

In [ ]:
cudf_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df['tconst'] = df['tconst']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df['averageRating'] = df['averageRating']

In [ ]:
%%time
## Transfer_post 0 ##
cudf_df['numVotes'] = df['numVotes']

In [ ]:
cudf_df2 = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 1 ##
cudf_df2['tconst'] = df2['tconst']

In [ ]:
cudf_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 1 ##
cudf_df['tconst'] = df['tconst']

In [4]:
%%time
### cell 1 ###

# Merging Rating and MetaData datsets into single dataframe w.r.t common column
main_df = pd.merge(df,df2,on='tconst')
main_df.head()

CPU times: user 8.5 s, sys: 448 ms, total: 8.95 s
Wall time: 8.92 s


,tconst,averageRating,numVotes,titleType,primaryTitle,originalTitle,isAdult,startYear,endYear,runtimeMinutes,genres
0,tt0000001,5.7,1914,short,Carmencita,Carmencita,0,1894,\N,1,\N
1,tt0000002,5.8,259,short,Le clown et ses chiens,Le clown et ses chiens,0,1892,\N,5,\N
2,tt0000003,6.5,1720,short,Pauvre Pierrot,Pauvre Pierrot,0,1892,\N,4,\N
3,tt0000004,5.6,172,short,Un bon bock,Un bon bock,0,1892,\N,12,\N
4,tt0000005,6.2,2537,short,Blacksmith Scene,Blacksmith Scene,0,1893,\N,1,\N


In [ ]:
cudf_main_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['tconst'] = main_df['tconst']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['averageRating'] = main_df['averageRating']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['numVotes'] = main_df['numVotes']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['titleType'] = main_df['titleType']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['primaryTitle'] = main_df['primaryTitle']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['originalTitle'] = main_df['originalTitle']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['isAdult'] = main_df['isAdult']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['startYear'] = main_df['startYear']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['endYear'] = main_df['endYear']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['runtimeMinutes'] = main_df['runtimeMinutes']

In [ ]:
%%time
## Transfer_post 1 ##
cudf_main_df['genres'] = main_df['genres']

**DATA Cleaning:**

In [ ]:
cudf_main_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['tconst'] = main_df['tconst']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['averageRating'] = main_df['averageRating']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['numVotes'] = main_df['numVotes']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['titleType'] = main_df['titleType']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['primaryTitle'] = main_df['primaryTitle']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['originalTitle'] = main_df['originalTitle']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['isAdult'] = main_df['isAdult']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['startYear'] = main_df['startYear']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['endYear'] = main_df['endYear']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['runtimeMinutes'] = main_df['runtimeMinutes']

In [ ]:
%%time
## Transfer_pre 2 ##
cudf_main_df['genres'] = main_df['genres']

In [5]:
%%time
### cell 2 ###

# creating a copy 
imdb_df = main_df.copy()

CPU times: user 330 ms, sys: 189 ms, total: 519 ms
Wall time: 518 ms


In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['tconst'] = imdb_df['tconst']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['averageRating'] = imdb_df['averageRating']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['numVotes'] = imdb_df['numVotes']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['titleType'] = imdb_df['titleType']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['primaryTitle'] = imdb_df['primaryTitle']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['originalTitle'] = imdb_df['originalTitle']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['isAdult'] = imdb_df['isAdult']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['startYear'] = imdb_df['startYear']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['endYear'] = imdb_df['endYear']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['runtimeMinutes'] = imdb_df['runtimeMinutes']

In [ ]:
%%time
## Transfer_post 2 ##
cudf_imdb_df['genres'] = imdb_df['genres']

**Currently, 'runtimeMinutes' column is in Minutes form thus we will convert it into Days:HH:MM:SS Format for better display of runtime information**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 3 ##
cudf_imdb_df['runtimeMinutes'] = imdb_df['runtimeMinutes']

In [6]:
%%time
### cell 3 ###

imdb_df['runtimeMinutes'].value_counts()

CPU times: user 57 ms, sys: 91 µs, total: 57.1 ms
Wall time: 56.1 ms


runtimeMinutes
\N      358359
30       42810
60       37284
22       36831
90       28662
         ...  
567          1
3900         1
2150         1
1620         1
468          1
Name: count, Length: 707, dtype: int64

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 4 ##
cudf_imdb_df['runtimeMinutes'] = imdb_df['runtimeMinutes']

In [7]:
%%time
### cell 4 ###

# Replace object type string '\N' with NaN First
imdb_df['runtime_delta'] = pd.to_numeric(imdb_df['runtimeMinutes'], errors='coerce', downcast="integer")

CPU times: user 582 ms, sys: 23.7 ms, total: 606 ms
Wall time: 602 ms


In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 4 ##
cudf_imdb_df['runtime_delta'] = imdb_df['runtime_delta']

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 5 ##
cudf_imdb_df['runtime_delta'] = imdb_df['runtime_delta']

In [8]:
%%time
### cell 5 ###

# Using 'pd.to_timedelta' to convert minutes into desired format
imdb_df['runtime_delta'] = pd.to_timedelta(imdb_df['runtime_delta'], unit='m')

CPU times: user 401 ms, sys: 8.02 ms, total: 409 ms
Wall time: 406 ms


In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 5 ##
cudf_imdb_df['runtime_delta'] = imdb_df['runtime_delta']

**Similarly, The 'isAdult' column should contain only two values, '0: non-adult title, 1: adult title' but here As we can see dataFrame column consist of mixed data values like string, bool and int. Thus, Lets sort and restructure this column into a Bool data type.**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 6 ##
cudf_imdb_df['isAdult'] = imdb_df['isAdult']

In [9]:
%%time
### cell 6 ###

imdb_df['isAdult'].unique()

CPU times: user 50 ms, sys: 4.2 ms, total: 54.2 ms
Wall time: 52.9 ms


array(['0', '1', '2020', '2014'], dtype=object)

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 7 ##
cudf_imdb_df['isAdult'] = imdb_df['isAdult']

In [10]:
%%time
### cell 7 ###

# Replace string and int by boolean
# we will leave out '2014' and '2020' as it is as there`s no way to confirm If they Adult or not not
imdb_df['isAdult'] = imdb_df['isAdult'].map({'0': 'Non_Adult_Title', '1': 'Adult_Title'})

CPU times: user 44.5 ms, sys: 3.95 ms, total: 48.5 ms
Wall time: 47.8 ms


In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 7 ##
cudf_imdb_df['isAdult'] = imdb_df['isAdult']

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 8 ##
cudf_imdb_df['isAdult'] = imdb_df['isAdult']

In [11]:
%%time
### cell 8 ###

imdb_df.isAdult = imdb_df.isAdult.fillna('Unrated')
imdb_df['isAdult'].unique()

CPU times: user 147 ms, sys: 7.88 ms, total: 155 ms
Wall time: 153 ms


array(['Non_Adult_Title', 'Adult_Title', 'Unrated'], dtype=object)

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 8 ##
cudf_imdb_df['isAdult'] = imdb_df['isAdult']

**Creating a new column to provide information of show run timeline in better way by joinning 'startYear' and 'endYear' columns into one**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 9 ##
cudf_imdb_df['startYear'] = imdb_df['startYear']

In [ ]:
%%time
## Transfer_pre 9 ##
cudf_imdb_df['endYear'] = imdb_df['endYear']

In [12]:
%%time
### cell 9 ###

# First, we should replace '\N' values to nan so next step
imdb_df['startYear'] = imdb_df['startYear'].replace('\\N', np.NaN)
imdb_df['endYear'] = imdb_df['endYear'].replace('\\N', np.NaN)

CPU times: user 189 ms, sys: 35 µs, total: 189 ms
Wall time: 188 ms


In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 9 ##
cudf_imdb_df['startYear'] = imdb_df['startYear']

In [ ]:
%%time
## Transfer_post 9 ##
cudf_imdb_df['endYear'] = imdb_df['endYear']

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 10 ##
cudf_imdb_df['startYear'] = imdb_df['startYear']

In [ ]:
%%time
## Transfer_pre 10 ##
cudf_imdb_df['endYear'] = imdb_df['endYear']

In [13]:
%%time
### cell 10 ###

# Now joinning them
# STEFANOS-DISABLE-FOR-MODIN
### ORIGINAL ###
# imdb_df['premiere_timeline'] = imdb_df[['startYear','endYear']].stack().groupby(level=0).agg('-'.join)
### COMPATIBLE WITH MODIN ###
# I tried a bunch of things, here's one:
#   imdb_df['premiere_timeline'] = imdb_df[['startYear','endYear']].stack().groupby(level=0).apply('-'.join)
# They don't seem to work. The problem seems to be with groupby(level=0). Generally, I'm not sure that Modin can
# groupby MultiIndex's.
# So, below I do the computation up to the stacking, just so that we can keep as much of it as possible (in case
# Modin, or some other system that uses the same benchmarks, can optimize it; we don't want to deprive this
# opportunity). Unfortunately, we cannot assign the result of stacking to premiere_timeline as they don't match.
# I just assign a dumb value to premiere_timeline so that it exists in the DF. It's not used later.

_ = imdb_df[['startYear','endYear']].stack()
imdb_df['premiere_timeline'] = 0

CPU times: user 164 ms, sys: 43.7 ms, total: 208 ms
Wall time: 206 ms


In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_post 10 ##
cudf_imdb_df['premiere_timeline'] = imdb_df['premiere_timeline']

**The data cleaning has been completed. However, Lets drop unneccessory columns from this dataframe for furthur analysis:**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['genres'] = imdb_df['genres']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['isAdult'] = imdb_df['isAdult']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['runtimeMinutes'] = imdb_df['runtimeMinutes']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['tconst'] = imdb_df['tconst']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['endYear'] = imdb_df['endYear']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['numVotes'] = imdb_df['numVotes']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['originalTitle'] = imdb_df['originalTitle']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['primaryTitle'] = imdb_df['primaryTitle']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['averageRating'] = imdb_df['averageRating']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['startYear'] = imdb_df['startYear']

In [ ]:
%%time
## Transfer_pre 11 ##
cudf_imdb_df['titleType'] = imdb_df['titleType']

In [14]:
%%time
### cell 11 ###

# Drop unneccessory columns from dataframe
columns_to_keep = ['tconst','averageRating','numVotes','titleType','primaryTitle',
                   'isAdult','premiere_timeline','runtime_delta']

imdb_df = imdb_df[columns_to_keep]

# Also renaming them with approriate name
renamed_cols = {'tconst':'IMDB_ID','averageRating':'Avg_Rating','numVotes':'Total_Votes',
                'titleType':'Title_Category','primaryTitle':'Title_Name','isAdult':'IN-18+'
                ,'premiere_timeline':'Air_time','runtime_delta':'Title_Runtime_Length'}

imdb_df.rename(columns = renamed_cols,inplace=True)
imdb_df.head()

CPU times: user 122 ms, sys: 8.08 ms, total: 130 ms
Wall time: 129 ms


,IMDB_ID,Avg_Rating,Total_Votes,Title_Category,Title_Name,IN-18+,Air_time,Title_Runtime_Length
0,tt0000001,5.7,1914,short,Carmencita,Non_Adult_Title,0,0 days 00:01:00
1,tt0000002,5.8,259,short,Le clown et ses chiens,Non_Adult_Title,0,0 days 00:05:00
2,tt0000003,6.5,1720,short,Pauvre Pierrot,Non_Adult_Title,0,0 days 00:04:00
3,tt0000004,5.6,172,short,Un bon bock,Non_Adult_Title,0,0 days 00:12:00
4,tt0000005,6.2,2537,short,Blacksmith Scene,Non_Adult_Title,0,0 days 00:01:00


**Top 10 highest voted and rated Titles across all categories**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 12 ##
cudf_imdb_df['Title_Name'] = imdb_df['Title_Name']

In [15]:
%%time
### cell 12 ###

top_10_highest = imdb_df.groupby(['Title_Name'])[
    ['Avg_Rating','Total_Votes']].max().sort_values(by=['Avg_Rating','Total_Votes'
                                                       ], ascending = False).reset_index().head(10)

top_10_highest

CPU times: user 2.56 s, sys: 3.7 ms, total: 2.57 s
Wall time: 2.56 s


,Title_Name,Avg_Rating,Total_Votes
0,Fight Club,10.0,2091999
1,Gladiator,10.0,1483822
2,Die Hard,10.0,869876
3,Ratatouille,10.0,727214
4,Limitless,10.0,570901
5,Shaun of the Dead,10.0,555530
6,Lost,10.0,549336
7,Fury,10.0,485732
8,Tangled,10.0,450335
9,Vertigo,10.0,400388


**Many Many other titles received the Perfect Rating of 10 across all Category but My personal favorite movie of all time, 'Fight Club' tops the table with Perfect Rating and highest Individual votes received while 'Gladiator' and 'Die Hard' takes the 2nd and 3rd spot respectivly on IMDB table.**

**Average votes and rating given to all categories**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 13 ##
cudf_imdb_df['Title_Category'] = imdb_df['Title_Category']

In [16]:
%%time
### cell 13 ###

highest_rated_type = imdb_df.groupby(['Title_Category'])[['Avg_Rating','Total_Votes']].mean(
).sort_values(by=['Avg_Rating','Total_Votes'], ascending = False).reset_index().round(decimals = 1)

highest_rated_type

CPU times: user 88.3 ms, sys: 7 µs, total: 88.3 ms
Wall time: 87.5 ms


,Title_Category,Avg_Rating,Total_Votes
0,tvEpisode,7.4,185.3
1,tvMiniSeries,7.1,1147.3
2,videoGame,6.9,357.4
3,tvShort,6.9,200.6
4,tvSeries,6.9,1423.6
5,short,6.8,72.4
6,tvSpecial,6.8,212.7
7,tvMovie,6.6,233.2
8,video,6.6,192.6
9,movie,6.2,3532.7


**Seems like 'tvEpisode' Category has received the best avg. Rating of 7.4/title when compaired to all other title types but received very less avg. votes per title. On the other hand, 'Movie' Category tops the chart as they received highest 3532.7 votes/titles on IMDB.**

**Category_wise longest running titles as per IMDB_database**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 14 ##
cudf_imdb_df['Title_Category'] = imdb_df['Title_Category']

In [ ]:
%%time
## Transfer_pre 14 ##
cudf_imdb_df['Title_Runtime_Length'] = imdb_df['Title_Runtime_Length']

In [17]:
%%time
### cell 14 ###

longest_runtime = imdb_df.loc[imdb_df.groupby(['Title_Category'], 
                                              sort=False)['Title_Runtime_Length'].idxmax()
                             ][['Title_Category', 'Title_Name', 'Title_Runtime_Length']]

longest_runtime.sort_values(by='Title_Runtime_Length', ascending = False).reset_index(drop=True)

CPU times: user 74.4 ms, sys: 0 ns, total: 74.4 ms
Wall time: 73.9 ms


,Title_Category,Title_Name,Title_Runtime_Length
0,movie,Logistics,35 days 17:00:00
1,video,The Longest Video on YouTube: 596.5 Hours,24 days 20:31:00
2,tvSpecial,Svalbard Minute by Minute,9 days 05:59:00
3,videoGame,Big Chungus,5 days 09:37:00
4,tvSeries,Great Chefs of the World,3 days 08:00:00
5,tvEpisode,The George Lucas Talk Show May the 4th Marathon,1 days 06:45:00
6,tvMiniSeries,War and Remembrance,1 days 03:00:00
7,tvMovie,24 Hours Berlin,1 days 00:00:00
8,short,Transsexuals from Space,0 days 01:53:00
9,tvShort,Prideland,0 days 00:55:00


**In our analysis, Movie named 'Logistics' has the longest run-time across all categories, at 35 Days and 17 Hours. While, 'The Longest Video on YouTube: 596.5 Hours' is in the second place.**

**Now, Lets check Category_wise shortest running titles as per IMDB_database**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 15 ##
cudf_imdb_df['Title_Category'] = imdb_df['Title_Category']

In [ ]:
%%time
## Transfer_pre 15 ##
cudf_imdb_df['Title_Runtime_Length'] = imdb_df['Title_Runtime_Length']

In [18]:
%%time
### cell 15 ###

shortest_runtime = imdb_df.loc[imdb_df.groupby(['Title_Category'], 
                                              sort=False)['Title_Runtime_Length'].idxmin()
                             ][['Title_Category', 'Title_Name', 'Title_Runtime_Length']]

shortest_runtime.reset_index(drop=True)

CPU times: user 75.1 ms, sys: 0 ns, total: 75.1 ms
Wall time: 74.7 ms


,Title_Category,Title_Name,Title_Runtime_Length
0,short,Awakening of Rip,0 days 00:00:00
1,movie,Fair at Ljutomer,0 days 00:01:00
2,tvEpisode,The Partridge Family,0 days 00:01:00
3,tvSeries,Beastly Behaviour,0 days 00:01:00
4,tvShort,The FA Cup Promotion,0 days 00:01:00
5,tvMovie,Das Bernstein-Amulett,0 days 00:02:00
6,tvMiniSeries,Loona TV,0 days 00:01:00
7,tvSpecial,Daisy,0 days 00:01:00
8,video,Sneeuwwitje,0 days 00:01:00
9,videoGame,Scud: The Disposable Assassin,0 days 00:02:00


**Short category film 'Awakening of Rip' holds the top place for being the shortest length title in IMDB_database.**

**Now, Lets shed some light on Average Ratings of Non_Adult Rated Titles vs Adult Rated Titles**

In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 16 ##
cudf_imdb_df['IN-18+'] = imdb_df['IN-18+']

In [ ]:
%%time
## Transfer_pre 16 ##
cudf_imdb_df['Title_Category'] = imdb_df['Title_Category']

In [ ]:
%%time
## Transfer_pre 16 ##
cudf_imdb_df['Avg_Rating'] = imdb_df['Avg_Rating']

In [19]:
%%time
### cell 16 ###

wanna_adult_or_not = imdb_df.groupby(['Title_Category','IN-18+']
                                    )['Avg_Rating'].agg('mean').round(1).unstack(fill_value= 0) 
wanna_adult_or_not.reset_index(inplace = True)
##### Visualise the chart in stack manner of bar type

# STEFANOS: Disable plotting
# plt.rcParams['figure.figsize'] = [15, 8]
# wanna_adult_or_not.plot(x='Title_Category', kind='bar', stacked=True, 
#                         title='Average Ratings of Non_Adult Rated Titles vs Adult Rated Titles')
# plt.xlabel('Title Categories')
# plt.ylabel('Average Ratings')
# plt.show()

CPU times: user 154 ms, sys: 16 ms, total: 170 ms
Wall time: 169 ms


In [ ]:
cudf_imdb_df = cudf.DataFrame()

In [ ]:
%%time
## Transfer_pre 17 ##
cudf_imdb_df['IN-18+'] = imdb_df['IN-18+']

In [ ]:
%%time
## Transfer_pre 17 ##
cudf_imdb_df['Title_Category'] = imdb_df['Title_Category']

In [ ]:
%%time
## Transfer_pre 17 ##
cudf_imdb_df['Avg_Rating'] = imdb_df['Avg_Rating']

In [20]:
%%time
### cell 17 ###

imdb_df.groupby(['Title_Category','IN-18+']
                                    )['Avg_Rating'].agg('mean').round(1).unstack(fill_value= 0) 

CPU times: user 163 ms, sys: 11.7 ms, total: 175 ms
Wall time: 173 ms


IN-18+,Adult_Title,Non_Adult_Title,Unrated
Title_Category,,,
movie,5.7,6.2,0.0
short,5.8,6.8,0.0
tvEpisode,7.0,7.4,7.4
tvMiniSeries,5.9,7.1,0.0
tvMovie,5.8,6.6,0.0
tvSeries,6.2,6.9,0.0
tvShort,6.3,6.9,0.0
tvSpecial,6.5,6.8,0.0
video,6.5,6.6,0.0


**'TvEpisode' category has received the highest Avg. Ratings in both Adult and Non-Adult rated title categories, However they there also the only title category among all whose some of the films/shows went Un-Rated.**

**Here ends this project, where we dig deep into the dataframe. First, We cleaned it and then sorted the data structure making a display of meaningful data. If you want to give any suggestion and point out mistakes please feel free to contact. Thanks!**